### PR TITLE
Add `cosign init` to initialize the SigStore root metadata

### PR DESCRIPTION
**Pull request by asraa**
* A command `cosign init` that initializes the local trusted SigStore root and pulls the trusted targets into `.sigstore/root`
* The initial root keys are added to `.sigstore/keys.json`. Clients need this to verify the root they pull
* The targets are used for the Fulcio cert and rekor public key, falling back for now to the previous embedded string in case `init` hasn't been run

Details:
* I implemented a "TUF remote store" located on a GCS bucket (for now it's a testing bucket I set up with the root metadata)
* We only download new metadata/targets from the remote location:
   - initially
   - if, when we retrieve a target, the local metadata no longer verifies
   - the local target does not match hash/size of the signed targets

Fixes
https://github.com/sigstore/cosign/issues/389

In practice, you can see if you run cosign without init'ing, you get log warnings asking "did you run `cosign init`?" otherwise, it's silent, we're using the local metadata.

```
asraa@asraa-glaptop:~/git/cosign$ COSIGN_EXPERIMENTAL=1 ./cosign verify gcr.io/asra-ali/base-builder-new
using embedded fulcio certificate. did you run `cosign init`? error retrieving target:  missing target metadata: tuf: no root keys found in local meta store
using embedded rekor public key. did you run `cosign init`? error retrieving target:  missing target metadata: tuf: no root keys found in local meta store

Verification for gcr.io/asra-ali/base-builder-new --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - Existence of the claims in the transparency log was verified offline
  - Any certificates were verified against the Fulcio roots.
Certificate subject:  [asraa@google.com]
{"critical":{"identity":{"docker-reference":"gcr.io/asra-ali/base-builder-new"},"image":{"docker-manifest-digest":"sha256:291750a9758506b9b3cace9c43e29b1a0f81172f43e9df1d293cb486a5d90abc"},"type":"cosign container image signature"},"optional":null}
asraa@asraa-glaptop:~/git/cosign$ ./cosign init
asraa@asraa-glaptop:~/git/cosign$ COSIGN_EXPERIMENTAL=1 ./cosign verify gcr.io/asra-ali/base-builder-new

Verification for gcr.io/asra-ali/base-builder-new --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - Existence of the claims in the transparency log was verified offline
  - Any certificates were verified against the Fulcio roots.
Certificate subject:  [asraa@google.com]
{"critical":{"identity":{"docker-reference":"gcr.io/asra-ali/base-builder-new"},"image":{"docker-manifest-digest":"sha256:291750a9758506b9b3cace9c43e29b1a0f81172f43e9df1d293cb486a5d90abc"},"type":"cosign container image signature"},"optional":null}

```

WIP because of tests and some defaults/globals I want to get rid of

### CODE REVIEW

**Comment by trishankatdatadog** [comment]
Hihi, would you need a review? Happy to help! 🙂

**Comment by asraa** [comment]
I have to do a bit of refactoring, but would love to get a review on the TUF code! I'll ping you

File: pkg/cosign/tlog.go [hunk 1]
```diff
@@ -41,6 +44,19 @@ import (
 // rekor.pub should be updated whenever the Rekor public key is rotated & the bundle annotation should be up-versioned
 //go:embed rekor.pub
 var rekorPub string
+var rekorTargetStr = `rekor.pub`
+
+func GetRekorPub() string {
```
Comment (dekkagaijin) [inline pkg/cosign/tlog.go:49]: IMHO we should put this somewhere under `cli/` in order to keep logging and `cosign` binary-specific out of package

File: pkg/cosign/tlog.go [hunk 2]
```diff
@@ -41,6 +44,19 @@ import (
...
+	ctx := context.Background() // TODO: pass in context?
+	buf := tuf.ByteDestination{Buffer: &bytes.Buffer{}}
+	err := tuf.GetTarget(ctx, rekorTargetStr, &buf)
+	if err != nil {
+		// The user may not have initialized the local root metadata. Log the error and use the embedded root.
+		fmt.Println("using embedded rekor public key. did you run `cosign init`? error retrieving target: ", err)
```
Comment (dekkagaijin) [inline pkg/cosign/tlog.go:55]:
```suggestion
		fmt.Fprintln(os.Stderr, "using embedded rekor public key. did you run `cosign init`? error retrieving target: ", err)
```

File: pkg/cosign/tuf/client.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+func Init(ctx context.Context) error {
+	rootClient, err := RootClient(ctx)
+	if err != nil {
+		return errors.Wrap(err, "initializing root client")
+	}
+	rootKeys, err := getRootKeys()
+	if err != nil {
+		return errors.Wrap(err, "retrieving root keys")
+	}
+	if err := rootClient.Init(rootKeys, TufThreshold); err != nil {
```
Comment (dekkagaijin) [inline pkg/cosign/tuf/client.go:138]: Pls make `TufThreshold` and `TufLocalTargets` args to init and/or provide a function that allows for non-default location and threshold

**Review by dekkagaijin — CHANGES_REQUESTED** [review]

File: pkg/cosign/tuf/client.go (hunk 3 above)
Comment (asraa) [reply pkg/cosign/tuf/client.go:138]: Yep!! Exactly the WIP part, these are gonna be args to `cosign init` with those defaults

File: pkg/cosign/tlog.go (hunk 1 above)
Comment (asraa) [reply pkg/cosign/tlog.go:49]: Ohhh hmmmm you mean keeping `fmt.Println` out? It's used elsewhere in this file as well

Comment (dekkagaijin) [reply pkg/cosign/tlog.go:49]: hrm yeah

File: pkg/cosign/tuf/client.go (hunk 3 above)
Comment (asraa) [reply pkg/cosign/tuf/client.go:138]: the only problem one is the location of the default local store:
* if you run `cosign init --local path/to/some/other/directory`
* and then you run `cosign verify`: how is it supposed to carry over to the non-default path?

i'm not convinced you should be able to configure that part, unless it's pulled from an env var

Comment (dekkagaijin) [reply pkg/cosign/tuf/client.go:138]: Pulling from an env var is pretty standard practice, e.g. `KUBECONFIG=` and `DOCKER_CONFIG=`

**Review by dekkagaijin — APPROVED** [review]
I'm fine with this for now, can always move packages around as we integrate more tuf stuff

**Comment by dekkagaijin** [comment]
@asraa squashing your local `go.sum` from upstream/main and running `go mod tidy` again should fix the conflict

**Comment by asraa** [comment]
All set, a local test added, and the path to the remote repository is in production.
@trishankatdatadog want to take a quick look at the TUF code? should be pretty simple.
Only question I had was: should we be making a `tuf.db` file? it works fine without one, except for in the test.

**Pull request merged**